torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -0,0 +1,106 @@
+from typing import List, NamedTuple, Optional
+
+import torch
+import torch.distributed as dist
+from torch.utils._contextlib import _DecoratorContextManager
+
+from ._fsdp_param import FSDPParam
+
+
+class AllGatherResult(NamedTuple):
+    all_gather_output: torch.Tensor
+    all_gather_event: Optional[torch.cuda.Event]
+    all_gather_work: Optional[dist.distributed_c10d.Work]
+    all_gather_input_numels: List[int]
+
+
+@torch.no_grad()
+def foreach_all_gather(
+    fsdp_params: List[FSDPParam],
+    group: dist.ProcessGroup,
+    async_op: bool,
+    all_gather_copy_in_stream: torch.cuda.Stream,
+    all_gather_stream: torch.cuda.Stream,
+    device: torch.device,
+    dtype: torch.dtype,
+) -> Optional[AllGatherResult]:
+    world_size, rank = (group.size(), group.rank())
+    # - Copy in
```

Comment:
I find PR description "Copy-in/All-Gather/Copy-Out Example"  to be very useful. maybe copy them to ``foreach_all_gather`` and ``foreach_all_gather_copy_out`` accordingly?